reactor: add background gate to wait on background tasks

bhalevy opened the pull request:
* reactor: add run_in_backround and close
  - run_in_background allows running background tasks safely, coupled with close() that wait on them using a gate.
* manual_clock: advance: use run_in_background to expire_timers
  - Fixes #1514

raphaelsc commented on include/seastar/core/reactor.hh:
```diff
@@ -359,6 +360,7 @@ private:
     bool _kernel_page_cache = false;
     std::atomic<bool> _dying{false};
+    gate _background_gate;
 private:
```
perhaps we could push this gate into a static thread_local member of manual_clock.

so manual_clock::advance() just needs to wrap its background task in with_gate(...).

and manual_clock::close() will wait on the gate, which is in turn called by reactor::close().

this way, we reduce the amount of changes to reactor.

I am not against adding this facility if it will be needed elsewhere. But perhaps we should consider the alternative I recommended if we're adding it only to handle a fragility in implementation detail of manual_clock.

bhalevy replied:
why limit the solution only to manual clock?
any other mechanism that would something alike will need to implement its own gate?
what's the point?

raphaelsc replied:
My point above is that we're very conservative on changes to reactor. If we really need it, ok.

bhalevy replied:
We have other background tasks like destroying of foreign ptrs that deserve this protection. It's a matter of fortunate timing that they finish running before the reactor exits.

bhalevy replied:
See https://github.com/scylladb/seastar/blob/1eea95e0201eb676f860fcf3b4faf4f03666b217/include/seastar/core/sharded.hh#L844-L859

bhalevy commented:
BTW, hit #1563 while modifying foreign_ptr_test.
It requires using smp::submit_to(0) to set_value on the promise while the `deferred` object is destroyed on shard 1.

michoecho reviewed (changes requested):
I'm not sure about the direction of this patch.

Since `run_in_background` will fail if called after the gate is closed, one has to ensure that it's always called before the gate is closed, and I don't think that's sustainable as a general framework for background tasks.

It seems to me that the only way to get a clear seastar shutdown is simply to keep running the main reactor loop until there is no more tasks. Otherwise things will leak, because stopping the reactor while there are still some tasks active can prevent some cleanups.
And if you do run the reactor until there are no more tasks, then you don't have to do any special work for background tasks. So doing special things for background tasks is misguided.

bhalevy commented on src/core/reactor.cc:
```diff
@@ -2965,6 +2965,35 @@ void reactor::add_urgent_task(task* t) noexcept {
+            return make_ready_future<>();
+        }
+        return engine()._background_gate.close();
+    });
```
makes sense

bhalevy replied:
made static in v2

bhalevy commented on src/core/reactor.cc:
```diff
@@ -2965,6 +2965,35 @@ void reactor::add_urgent_task(task* t) noexcept {
+            std::rethrow_exception(f.get_exception());
+        }
+        return;
+    }
```
It throws if gate is closed. I'm not sure we want to disrupt the application in this case since by not waiting on the background fiber the caller indicates they don't care if it succeeds or not.
Otherwise, they must wait on it.

bhalevy commented on src/core/reactor.cc:
```diff
@@ -2965,6 +2967,35 @@ void reactor::add_urgent_task(task* t) noexcept {
+future<> reactor::close() {
+    seastar_logger.debug("reactor::close");
+    return smp::invoke_on_all([] {
+        if (engine()._background_gate.is_closed()) {
```
it depends, we generally don't but sometimes we do (even reactor::exit can be called multiple times the way it's written)

bhalevy replied:
renamed to `drain()` and still allowed to be called more than once (it makes more sense for drain than for close)

bhalevy commented on src/core/reactor.cc:
```diff
@@ -2965,6 +2967,35 @@ void reactor::add_urgent_task(task* t) noexcept {
+void reactor::run_in_background(future<> f) {
+    if (f.available()) {
+        if (f.failed()) {
+            std::rethrow_exception(f.get_exception());
```
once scheduled, it can no longer return the error, just log that it happened and wasn't handled.
it's debatable if we want to throw if the gate is closed.

bhalevy replied:
this part removed in v2, so the error will be handled (and ignored) as if it was asynchronous

bhalevy commented on src/core/app-template.cc:
```diff
@@ -155,7 +155,11 @@ int
+            return func_done->get_future().finally([] {
+                return engine().close();
+            });
+        });
```
I want it to run last after the `at_exit` tasks.
Maybe the reactor should take care of that on its own by running close/drain in `run_exit_tasks()`.

bhalevy replied:
call to `drain()` moved to engine::stop so that the app framework doesn't have to do anything

bhalevy commented on src/core/reactor.cc:
```diff
@@ -2965,6 +2967,35 @@ void reactor::add_urgent_task(task* t) noexcept {
+        return;
+    }
+    try {
+        // _backgroud_gate closed in reactor::close()
```
fixed in v2

bhalevy commented on include/seastar/core/reactor.hh:
```diff
@@ -564,6 +566,16 @@ public:
+    }
+
+    // Waits for all background tasks
+    future<> close();
```
changed to `drain` in v2

bhalevy commented:
In v2 ([0e53d6f](https://github.com/scylladb/seastar/commit/0e53d6feb92edcb0d542286114d17dcbeacf43bf)):
- close() renamed to drain()
  - made static
  - automatically called from engine::stop (on shard 0 only)
- fixed typo in comment
- do not special handle synchronous failed future in run_in_background()
- removed some left over log messages in unit test

bhalevy commented:
ping @tgrabiec @avikivity 
I think this PR offers net gain to stability
even though it's not perfect.
Please consider merging

tgrabiec commented on include/seastar/core/sharded.hh:
```diff
@@ -851,12 +852,14 @@ private:
     static future<> destroy_on(PtrType p, unsigned cpu) noexcept {
         if (p) {
             if (cpu != this_shard_id()) {
+              engine().run_in_background([cpu, p = std::move(p)] () mutable {
```
This is not a background task, the future is (was) returned to the caller. 

Now this always returns a ready future, which is wrong.

bhalevy commented:
In v3 ([6c38001](https://github.com/scylladb/seastar/commit/6c38001afa412f4525b9b29fb103929790159d5b))
- foreign_ptr: destroy: run_in_background only when needed on the synchronous destroy path.
